feat: Add user statistics to home screen

haroldadmin opened the pull request:
Fixes issue systers#20. This adds user stats to the home screen
containing numbers for relations/requests, and the most recent
achievements.

### Description
Using the new /home endpoint on the backend, this PR adds user statistics to the home fragment.
It displays the count of various types of relations/requests, as well as recent achievements.

Fixes #20 

I have used the SingleLiveEvent class in this PR to show snackbars on the screen in case of an error. I picked this approach because of the following reasons:
1. Exposing the snackbar messages as an Observable is a clean way of separating responsibilites between the viewmodel and the view.
2. The SingleLiveEvent class ensures that the snackbar message is shown only once. If a regular LiveData object were used here, then the view would be notified of the latest value in the message observable on every configuration change. So if the user receives a snackbar message, they would receive it again if they rotate the scren.

Here's a demonstration of the SingleLiveEvent class in action: 
![Demonstration](https://user-images.githubusercontent.com/24315306/52612721-ee618100-2eb0-11e9-95e5-a70dd32df1b9.gif)

As you can see in the gif, the homescreen uses the SingleLiveEvent class to display Snackbars and therefore it is not shown again on configuration change, whereas the Profile Fragment does not use SLE and therefore its Snackbars are shown again on config change.

I have translated the SingleLiveEvent [source](https://github.com/googlesamples/android-architecture/blob/dev-todo-mvvm-live/todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/SingleLiveEvent.java) from Java to Kotlin in order to maintain a 100% Kotlin codebase :)

### Type of Change:
- Code
- User Interface

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
Tested on physical device and emulator.

<img src="https://user-images.githubusercontent.com/24315306/50726949-06b7e080-113a-11e9-9dd6-ff1d2929b2eb.jpeg" width="45%"></img> <img src="https://user-images.githubusercontent.com/24315306/50726950-091a3a80-113a-11e9-8159-3b7d7f00c471.jpeg" width="45%"></img> 


### Checklist:
**Delete irrelevant options.**

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] Any dependent changes have been merged

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes

mohak1283 commented:
Nice work @haroldadmin :+1: I saw your code and it looks great to me :)

haroldadmin commented:
Thank you for the review @mohak1283! Is there anything you liked in particular?

mohak1283 commented:
> > Nice work @haroldadmin +1 I saw your code and it looks great to me :)
> 
> Thank you for the review Mohak! Is there anything you liked in particular?

Yeah that SingleLiveEvent.kt class :+1: 
Can you give the overview of that class like for what purpose it is created for? I read the comments but little bit confused.

haroldadmin commented:
@mohak1283 SingleLiveEvent is actually from Google's Architecture Samples. Here's a nice article explaining their use case: [LiveData with SnackBar, Navigation and other events (the SingleLiveEvent case)](https://medium.com/androiddevelopers/livedata-with-snackbar-navigation-and-other-events-the-singleliveevent-case-ac2622673150)

isabelcosta commented on app/src/main/java/org/systers/mentorship/view/fragments/HomeFragment.kt:
```diff
@@ -17,4 +35,47 @@ class HomeFragment : BaseFragment() {
+        with(homeViewModel) {
+            userStats.observe(viewLifecycleOwner, Observer { stats ->
+                binding.stats = stats
+                if (stats?.achievements?.isEmpty() != false) {
```
This condition confuses me. `isEmpty()` function already returns a boolean, you can remove `!= false` inside if, and work with that logic.

haroldadmin replied:
Initially I too did it the way you suggest, but Android Studio said that that way could cause a problem when the `stats` variable becomes null.

Let's look at it this way: if `stats` or `stats.achievements` is null, then the condition inside the if statement evaluates to null. An if statement requires a `Boolean`, but if we do things this way we would be passing it a `Boolean?`. To fix that, we can either add a non-null assertion `!!` operator to force the compiler to treat this expression as a non null Boolean, or we can use the elvis operator `?:` to supply an alternate value in case the expression evaluates to null.

Since using `!!` is unsafe, I used the elvis operator and my if statement looked this way:

`if (stats?.achievements?.isEmpty() ?: false)`

But when I do this, Android Studio suggests this:

`Equality check should be used instead of elvis for nullable boolean check`

And so we have an if statement that looks like this:

`if (stats?.achievements?.isEmpty() != false)`

isabelcosta commented on app/src/main/java/org/systers/mentorship/view/adapters/AchievementsAdapter.kt:
```diff
@@ -0,0 +1,60 @@
+
+        /**
+         * This function binds the description of the achievement to the textview
+         *
```
be consistent please. in the class KotlinDocs, you don't have a space between the two lines of text. And here you have. Can you remove it here maybe, or add a line in the KotlinDocs above

haroldadmin replied:
I have removed the blank line now. Thank you for pointing this out. Consistent documentation is important.

isabelcosta commented on app/src/main/java/org/systers/mentorship/view/adapters/AchievementsAdapter.kt:
```diff
@@ -0,0 +1,60 @@
+import kotlinx.android.synthetic.main.achievement_list_item.view.*
+import org.systers.mentorship.R
+import org.systers.mentorship.models.Task
+
```
remove extra line

haroldadmin replied:
Whoops. I've removed it now.

isabelcosta commented on app/src/main/java/org/systers/mentorship/remote/datamanager/UserDataManager.kt:
```diff
@@ -43,4 +44,13 @@ class UserDataManager {
     }
+
+    /**
+     * This will call the getHomeStats method of the UserService
```
This can be seen by your code, you can just simply say: `This function fetches user statistics`

haroldadmin replied:
Okay, I've changed this.

isabelcosta commented on app/src/main/java/org/systers/mentorship/models/Task.kt:
```diff
@@ -0,0 +1,34 @@
+ *
+ * @param id The id of this task
+ * @param description The description of this task
+ * @param isDone Represents whether this task has been
```
`completed` word is in the line below, unnecessarily, right?
Can you fix that in the lines below

haroldadmin replied:
Yep, fixed now.

isabelcosta commented on app/src/main/java/org/systers/mentorship/models/HomeStatistics.kt:
```diff
@@ -0,0 +1,39 @@
+import kotlinx.android.parcel.Parcelize
+
+/**
+ * This data class represents the response of the /home endpoint
```
You don't need to reference the endpoint. You never know if this can be reused for something else.
Say:
```
This class represents statistics of the user actions on the app.
```

haroldadmin replied:
Okay, fixed now. I've also removed the unnecessary blank line below for consistency.

isabelcosta commented on app/src/main/java/org/systers/mentorship/remote/services/UserService.kt:
```diff
@@ -48,4 +49,12 @@ interface UserService {
     fun updateUser(@Body user: User): Observable<CustomResponse>
 
+    /**
+     * This function gets the current user's home screen stats
```
stats -> statistics

haroldadmin replied:
Fixed now.

isabelcosta commented on app/src/main/res/layout/fragment_home.xml:
```diff
@@ -1,13 +1,194 @@
+            android:animateLayoutChanges="true"
+            tools:context=".view.fragments.HomeFragment">
+
+            <TextView
```
Can you use [AppCompatTextView](https://developer.android.com/reference/kotlin/androidx/appcompat/widget/AppCompatTextView) in your text views?

haroldadmin replied:
Okay, I've changed it too AppCompatTextview

isabelcosta commented on app/src/main/java/org/systers/mentorship/view/adapters/AchievementsAdapter.kt:
```diff
@@ -0,0 +1,60 @@
+
+
+/**
+ * This class is the recyclerview adapter for achievements
```
this can be 2 lines instead of 4 lines of text right?

haroldadmin replied:
I can bring it down to 3, because Android Studio by default wraps lines at 100 characters. Here's how it looks like now:

```
/**
 * This class is the recyclerview adapter for achievements. It is a subclass of [ListAdapter] for
 * easy async calculation of diffs using DiffUtil to provide nice animations when the data set
 * changes.
 */
```

isabelcosta commented on app/src/main/java/org/systers/mentorship/view/adapters/AchievementsAdapter.kt:
```diff
@@ -0,0 +1,60 @@
+ * This class represents the DiffCallback
+ * for tasks, and is used in [AchievementsAdapter]
+ */
+class AchievementsItemCallback : DiffUtil.ItemCallback<Task>() {
```
I don't see where you are using this class and its functions. If this is from an old implementation of your's please remove this.

haroldadmin replied:
This is the DiffUtil.ItemCallback implementation for the Task model class, and it is used in the achievements in the `AchievementsAdapter` class. It is a required parameter in the constructor. 

`class AchievementsAdapter : ListAdapter<Task, AchievementsAdapter.ViewHolder>(AchievementsItemCallback())`

ListAdapter requires an implementation like this to be passed to the constructor.

isabelcosta commented on app/src/main/java/org/systers/mentorship/viewmodels/HomeViewModel.kt:
```diff
@@ -0,0 +1,79 @@
+    val message: LiveData<String>
+        get() = _message
+
+    init {
```
Can you explain why call the function inside `init {}` instead of a function, like other ViewModels? @haroldadmin

haroldadmin replied:
Yes, sure. I had thought about its implementation in other ways, but I ultimately landed on this one because:

The home fragment is the first fragment that is visible to the users when they open the app. This network call should therefore be made as early as possible to ensure that the data is fetched and displayed at the earliest possible time. Therefore, instead of waiting for the Fragment to complete its initialization, I make the network call as soon as the ViewModel is constructed. This way, we can save some time.

I hope this makes it a little clearer, but if you still suggest I move the call from an init block to a function, I can do that too. What do you think?

isabelcosta commented on app/src/main/java/org/systers/mentorship/viewmodels/HomeViewModel.kt:
```diff
@@ -0,0 +1,79 @@
+    private val compositeDisposable by lazy { CompositeDisposable() }
+
+    private val _userStats = MutableLiveData<HomeStatistics>()
+    private val _message = SingleLiveEvent<String>()
```
Do you really need the SingleLiveEvent? Doesn't a fragment keep the configuration, and avoid the problems you mention as a reason to use SingleLiveEvent class? I used this in the past, but I can't' remember why I stopped using it.
cc @m-murad

haroldadmin replied:
I think fragments do keep some of their state, but not their view state. View hierarchy is destroyed and recreated upon config changes.

My reasoning for using SingleLiveEvent instead of regular LiveData was as follows:
1. This live data is observed using `viewLifecycleOwner` as the lifecycle owner in the fragment, so when the view hierarchy is destroyed this subscription will stop.
2. In the event of a configuration change, the view hierarchy is formed again, and so this subscription will be made again.
3. As is LiveData's nature, it will again give me the latest value inside the LiveData upon re-subscription.
4. Therefore, upon config change, I'll re-receive the Snackbar message and the Snackbar will be shown again to the user if I'm not using SingleLiveEvent.

So the question arises, why use `viewLifecycleOwner` instead of `this` as the lifecycle owner? The reason is that since a fragment does indeed keep its state, it is not destroyed upon config change, and the subscription is never terminated. Upon recreation of the view, a new subscription is made without cancelling the previous one. This issue was brought up in some keynote at this year's Google IO.

I hope this makes things clearer. What do you think @isabelcosta?

m-murad replied:
When `LifecycleOwner` is destroyed for recreation, `ViewModel` is preserved. Read [this](https://developer.android.com/topic/libraries/architecture/viewmodel)

haroldadmin replied:
@m-murad I know this. Since the ViewModel is preserved, the LiveData objects we expose from it can hold on to their value. However, when the fragment is recreated and it re-subscribes to these live data objects, it get's updated with their latest values again. This is the expected behaviour in most cases, but when we are using LiveData for snackbars or toasts, then they also appear again on the screen. This is why SingleLiveEvent was created, right?

isabelcosta reviewed (dismissed):
Thank you for your contribution @haroldadmin 🎉 
I requested some changes for you to address and some questions about your approach.
Please address them (when you can).

haroldadmin commented:
@isabelcosta I have added the requested changes. I have also added explanations on some of my implementation methods in the issues that you pointed out, so please take a look at those too!

m-murad commented on app/src/main/java/org/systers/mentorship/models/HomeStatistics.kt:
```diff
@@ -0,0 +1,36 @@
+@SuppressLint("ParcelCreator")
+@Parcelize
+data class HomeStatistics(
+        @SerializedName("name")
```
When `variable_name` and `json_name` are same, don't use `@SerializedName` annotation

haroldadmin replied:
While serialized name is not needed in this case, I still think its always a good idea to keep it around because in the future if the API changes and returns variables with a new name, we would just need to change its serialized name while the rest of our code can stay the same.

I don't know if this is a best practice to follow or not. I do agree that this adds a little complexity unnecessarily right now. I'll remove it since you request it.

m-murad commented on app/src/main/java/org/systers/mentorship/models/HomeStatistics.kt:
```diff
@@ -0,0 +1,36 @@
+        val cancelledRelations: Int,
+        @SerializedName("rejected_requests")
+        val rejectedRequests: Int,
+        @SerializedName("achievements")
```
Remove `@SerializedName`

haroldadmin replied:
Okay.

m-murad commented on app/src/main/java/org/systers/mentorship/utils/SingleLiveEvent.kt:
```diff
@@ -0,0 +1,59 @@
+package org.systers.mentorship.utils
```
What have you used this for?

haroldadmin replied:
I've used it to notify the user with a snackbar when the API call for home statistics fails. It's best to use SingleLiveEvent instead of LiveData for this, because it makes sure that Snackbar/Toast does not reappear on the screen in the event of a configuration change.

I got the idea here:
[LiveData with SnackBar, Navigation and other events (the SingleLiveEvent case)](https://medium.com/androiddevelopers/livedata-with-snackbar-navigation-and-other-events-the-singleliveevent-case-ac2622673150)

m-murad commented on app/src/main/java/org/systers/mentorship/view/adapters/AchievementsAdapter.kt:
```diff
@@ -0,0 +1,56 @@
+/**
+ * This class represents the DiffCallback for tasks, and is used in [AchievementsAdapter]
+ */
+class AchievementsItemCallback : DiffUtil.ItemCallback<Task>() {
```
What are these callbacks for?

haroldadmin replied:
This callback enables the use of DiffUtil on the RecyclerView for Achievements on the home screen. Using this callback we can get smooth animations when the underlying dataset of the RecyclerView Adapter changes. It's not necessary to use DiffUtil with any RecyclerView, but I thought it would be nice to have it. It does improve the peformance though, because it is more efficient than calling `notifyDatasetChanged()`.

m-murad commented on app/src/main/java/org/systers/mentorship/models/HomeStatistics.kt:
```diff
@@ -0,0 +1,36 @@
+ * @param achievements a list of upto 3 completed tasks
+ */
+
+@SuppressLint("ParcelCreator")
```
Remove this. Update your IDE. This bug has been fixed

haroldadmin replied:
I added this annotation because Travis build failed without. The lint check on Travis failed without this annotation with this error:

`This class implements Parcelable but does not provide a CREATOR field [ParcelCreator]`

haroldadmin commented:
@isabelcosta @m-murad I have added the requested changes. Please review when you can.

haroldadmin commented:
@isabelcosta @m-murad I have added the requested changes. I've removed the `@SerializedName` annotation where it was not necessary. Please review when you can.

haroldadmin commented:
@isabelcosta I've resolved the merge conflict. The PR is up to date now.

isabelcosta reviewed (commented):
I might have to run this and test it on my device, then I can give you a proper approval @haroldadmin

haroldadmin commented:
> I might have to run this and test it on my device, then I can give you a proper approval @haroldadmin

Okay, if you want I can provide you with a built APK so that you don't have to build it yourself. Let me know if you need it, I'll send it to you on Slack.

isabelcosta reviewed (dismissed):
@haroldadmin I ran this on my device, and it looks good, but I would like another approval on this issue before merging.

haroldadmin commented:
@m-murad Can you give your approval to this PR?

isabelcosta reviewed (commented):
@haroldadmin fix merge conflicts please

haroldadmin committed:
feat: Add user statistics to home screen

Fixes issue #20. This adds user stats to the home screen
containing numbers for relations/requests, and the most recent
achievements.

haroldadmin commented:
@isabelcosta I've resolved the merge conflicts now.

haroldadmin commented:
@m-murad @isabelcosta I am not sure what is the current status of this PR. I have incorporated all the changes that were requested. Merge conflicts have cropped up a couple of times due to inactivity. Please review and merge, if everything looks good.

isabelcosta reviewed (approved):
Thank you for your contribution @haroldadmin !
This was a big PR to review, but ultimately I think it does what is supposed to.
